crates/rpc/src/jsonrpc/error.rs

```diff
@@ -61,20 +48,36 @@ impl RpcError {
             RpcError::InvalidParams => "Invalid params".into(),
             // TODO: this is not necessarily a good idea. All internal errors are returned here, even
             // ones that we probably should not disclose.
-            RpcError::InternalError(e) => e.to_string().into(),
-            RpcError::ApplicationError { message, .. } => message.into(),
-            RpcError::WebsocketSubscriptionClosed { reason, .. } => {
-                ("Websocket subscription closed, reason: ".to_string() + reason).into()
-            }
+            RpcError::InternalError(_) => "Internal error".into(),
+            RpcError::ApplicationError(e) => e.to_string().into(),
+            RpcError::WebsocketSubscriptionClosed { .. } => "Websocket subscription closed".into(),
         }
     }
 
     pub fn data(&self) -> Option<Value> {
         match self {
             RpcError::WebsocketSubscriptionClosed {
-                subscription_id, ..
-            } => Some(Value::Number(Number::from(*subscription_id))),
-            _ => None,
+                subscription_id,
+                reason,
+            } => Some(json!({
+                "id": subscription_id,
+                "reason": reason,
+            })),
+            RpcError::ApplicationError(e) => e.data(),
+            RpcError::InternalError(error) => {
+                let error = error.to_string();
+                if error.is_empty() {
+                    None
+                } else {
+                    Some(json!({
+                        "error": error.to_string(),
+                    }))
```

Comment:
I feel obligated to mention security concerns here, but at the same time this is open source software.
Are we sure we don't expose unnecessary details here?
Would operators care about exposing potentially security-sensitive data?

Reply:
Alternatively we could add a public error message for internal errors, giving us the opportunity to review whether we expose sensitive data.

Reply:
Its a good question and something I do worry about as well. I think at this top level we probably should suppress this error - but maybe log it if encountered?

There is also the application error internal error which I think sometimes gets used to communicate actual information -- which is probably a bad idea as well?

Reply:
I think whitelist approach would be necessary here, with a public error message field as a possible implementation

Reply:
> There is also the application error internal error which I think sometimes gets used to communicate actual information -- which is probably a bad idea as well?

It is a bad idea. The underlying issue is that for _some_ errors there's no proper error code in the specification but we still want to return the information.

What we could do is introducing a `Custom` error  instead of using `Internal` and making sure that all such cases now use `Custom`.

Reply:
On the other hand -- that change should definitely be in a new PR. We've been returning `Internal` error details pretty much forever so this PR does not change that.

Reply:
I think that is the correct idea yeah. I'll attempt to do this:

1. No `data` for internal errors
2. Log internal errors
3. `Custom` for specific errors